src/v/cluster/partition_allocator.cc

```diff
@@ -64,34 +68,79 @@ static bool is_machine_in_replicas(
 }
 
 std::optional<std::vector<model::broker_shard>>
-partition_allocator::allocate_replicas(int16_t replication_factor) {
-    std::vector<model::broker_shard> replicas;
-    replicas.reserve(replication_factor);
+partition_allocator::allocate_replicas(
+  int16_t replication_factor, const std::vector<model::broker_shard>& current) {
+    std::vector<model::broker_shard> new_replicas;
+    new_replicas.reserve(replication_factor);
+    for (size_t i = 0; i < _machines.size(); ++i) {
+        const uint16_t replicas_left = replication_factor
+                                       - (new_replicas.size() + current.size());
+        if (replicas_left == 0) {
+            break;
+        }
 
-    while (replicas.size() < (size_t)replication_factor) {
-        const uint16_t replicas_left = replication_factor - replicas.size();
         if (_available_machines.size() < replicas_left) {
-            rollback(replicas);
+            rollback(new_replicas);
             return std::nullopt;
         }
         auto& rr = round_robin_ptr();
         auto& machine = *rr;
         rr++;
-        if (is_machine_in_replicas(machine, replicas)) {
+        if (machine.is_decommissioned()) {
```

Comment:
In general I think we should avoid this situation. I didn't add any validation preventing doing this. In general it may happen that node running new redpanda version may crash with `vassert`. Do you think we should add verification to check if all nodes in a cluster support decommissioning to check if we are able to decommission a node ?